explorer/server.go

```diff
@@ -0,0 +1,209 @@
+	}
+}
+
+func (a *App) getNode(w http.ResponseWriter, r *http.Request) {
+
+	nodeID := mux.Vars(r)["node_id"]
+	value, _ := a.GetRedisKey(fmt.Sprintf("GRID3NODE:%s", nodeID))
+
+	// No value, fetch data from the node
+	if value == "" {
+		nodeInfo, err := a.fetchNodeData(r.Context(), nodeID)
+		if err != nil {
+			log.Error().Err(err).Msg("could not fetch node data")
+			w.WriteHeader(http.StatusBadRequest)
+			w.Write([]byte(http.StatusText(http.StatusBadRequest)))
+			return
+		}
+		// Save value in redis
+		// caching for 30 mins
+		marshalledInfo, err := json.Marshal(nodeInfo)
+		if err != nil {
+			log.Error().Err(err).Msg("could not marshal node info")
+			w.WriteHeader(http.StatusInternalServerError)
+			w.Write([]byte(http.StatusText(http.StatusInternalServerError)))
+			return
+		}
+		err = a.SetRedisKey(fmt.Sprintf("GRID3NODE:%s", nodeID), marshalledInfo, 30*60)
+		if err != nil {
+			log.Warn().Err(err).Msg("could not cache data in redis")
+		}
+		value = string(marshalledInfo)
+	}
+
+	w.WriteHeader(http.StatusOK)
+	w.Header().Add("Content-Type", "application/json")
+	w.Write([]byte(value))
+}
+
+func (a *App) fetchNodeData(ctx context.Context, nodeID string) (NodeInfo, error) {
+	twinID, err := getNodeTwinID(nodeID, a.explorer)
+	if err != nil {
+		return NodeInfo{}, errors.Wrap(err, "could not get node twin ID")
+
+	}
+
+	nodeClient := client.NewNodeClient(twinID, a.rmb)
+	NodeCapacity, UsedCapacity, err := nodeClient.Counters(a.ctx)
+	if err != nil {
+		return NodeInfo{}, errors.Wrap(err, "could not get node capacity")
+	}
+
+	dmi, err := nodeClient.SystemDMI(ctx)
+	if err != nil {
+		return NodeInfo{}, errors.Wrap(err, "could not get node DMI info")
+	}
+
+	hypervisor, err := nodeClient.SystemHypervisor(ctx)
+	if err != nil {
+		return NodeInfo{}, errors.Wrap(err, "could not get node hypervisor info")
+	}
+
+	capacity := capacityResult{}
+	capacity.Total = NodeCapacity
+	capacity.Used = UsedCapacity
+
+	return NodeInfo{
+		Capacity:   capacity,
+		DMI:        dmi,
+		Hypervisor: hypervisor,
+	}, nil
+
+}
+
+func (a *App) runServer(hostAddress string) {
+	log.Info().Str("listening on", hostAddress).Str("explorer", a.explorer).Msg("Server started ...")
+}
+
+// Setup is the server and do initial configurations
+func Setup(router *mux.Router, debug bool, explorer string, redisServer string, hostAddress string) {
+	log.Info().Str("redis address", redisServer).Msg("Preparing Redis Pool ...")
+
+	redis := &redis.Pool{
+		MaxIdle:   10,
+		MaxActive: 10,
+		Dial: func() (redis.Conn, error) {
+			conn, err := redis.Dial("tcp", redisServer)
+			if err != nil {
+				log.Error().Err(err).Msg("fail init redis")
+			}
+			return conn, err
+		},
+	}
+
+	rmbClient, err := rmb.Default()
+	if err != nil {
+		log.Error().Err(err).Msg("couldn't connect to rmb")
+		return
+	}
+
+	a := App{
+		debug:    debug,
+		explorer: explorer,
+		redis:    redis,
+		ctx:      context.Background(),
+		rmb:      rmbClient,
+	}
+	go a.runServer(hostAddress)
```

Comment:
this just logs ?

Reply:
nothing for now, for the future import go-rmb and start its server